Create CrossCompilationModel.md (#73581)
swiftlang/swift: docs/CrossCompilationModel.md

Reviewer: Could we put this file in a `/docs/CrossCompilation` or similar directory for better organization? Also, please consider adding a mapping for the new file/directory to `/.github/CODEOWNERS`.

Author: I don't think that the proposed organisation is that much better - ultimately, this is about how the compiler works, what the flags do, and how they enable cross-compilation. I'm happy to rename the file though if you have a suggestion.

inputs outside of the user provided code.

1. Compiler Resources
2. System (C/C++) Headers (Platform SDK)
3. Platform Swift Modules (Swift SDK)

These pieces compose in a particular manner to build a system image to build
code against.

The compiler resources are content shipped with the toolchain and are tied to
that specific version and build of the compiler. In the case of the Swift
compiler, this includes the Swift shims. Whilst this is a compiler resource, the
packaging may not necessarily be part of the toolchain due to interdependencies.
Some of this content is required to process the system headers themselves (e.g.
clang's builtin resource headers).

The C/C++ system headers (and libraries) are what is traditionally called the
Unix "sysroot". On most Unix systems, this is the set of headers that are
associated with the system library set which normally may be found in
`/usr/include`. On Darwin, this is included in the unified SDK which is shipped
to the user, while on Windows, this is called the Windows SDK, which is a
separate installable component. For simplicity, we will refer to this as the
Platform SDK universally.

> The platform Swift modules provided by the Swift SDK is unified with the

Reviewer 2: Maybe it's just me, but my brain is having a hard time parsing this sentence.

> The platform Swift modules provided by the Swift SDK is unified with the Platform SDK.

Is this that the module maps for the platform are shipped as part of the Swift SDK? I think that is already covered by

>  It contains the libraries and overlays that provide the core language runtime and expose system libraries to Swift code in an ergonomic manner. This may be in the form of API Notes, module maps, wrapper types, or Swift `extension`s, or entire libraries.

I would like to see Swift SDK defined before use though.
Maybe something more like this?

> The Swift SDK contains libraries and overlays that provide the core language
runtime and expose system libraries to Swift code in an ergonomic manner. This
may be in the form of API Notes, module maps, wrapper types, Swift
`extension`s to platform types, or entire overlay libraries. In some distributions, the Swift SDK and
platform SDK are unified, such as the SDKs for Apple platforms, where the C, C++,
and Swift libraries and interfaces are shipped together.

Thoughts?

Author: Adjusted as discussed offline.

> # Crosss-Compilation Model

Reviewer 3: ```suggestion
# Cross-Compilation Model
```
Doing my part as spellchecker, again😅

## Components

When compiling Swift code, the compiler will consult three different sources of
inputs outside of the user provided code.

1. Compiler Resources
2. System (C/C++) Headers
3. Platform Swift Modules (SDK Overlay)

These pieces compose in a particular manner to build a system image to build
code against.

> The compiler resources are content for the compiler that the compiler provides.

Author: Yeah, but also there is the compatibility required for the standard library. The Swift standard library depends on some of the compiler resources. The reason for calling out the packging here is that at least on Windows, we package the SwiftShims into the SDK to maintain the compatibility with the standard library. But I agree with you that we should make this bit explicit.

> The compiler resources are content for the compiler that the compiler provides.

Reviewer 2: Yeah... SwiftShims is unfortunate since it combines headers that are definitely compiler-ABI details with headers that are for the stdlib. So for that one, you can flip a coin to choose whether it's compiler-specific or stdlib-specific and you'd be equally right and wrong at the same time.

> In the case of the Swift compiler, this includes the Swift shims. Whilst this is

Author: Standard library macros are new! However, I think that they should be going into the SDK - they are part of the standard library which goes into the SDK.

> In the case of the Swift compiler, this includes the Swift shims. Whilst this is

Reviewer 2: Yes... but.... what triple do we compile them for?
I'm being somewhat facetious here.

> In the case of the Swift compiler, this includes the Swift shims. Whilst this is

Author: I think that the standard library macros _should_ go into the SDK, which is technically a change and something we should try to get in for 6.0.

a compiler resource, the packaging may not be necessarily part of the toolchain
due to interdependencies. Some of this content may be required to process the
system headers themselves (e.g. clang's builtin resource headers).

The C/C++ system headers (and libraries) are what is traditionally called the
Unix "sysroot". On Darwin, this is the SDK which are shipped to the user, while
on Windows, this is called the Windows SDK, which is a separate installable
component.

The platform Swift modules provided by the SDK overlay overlays the sysroot or
(C/C++) SDK. This overlay is a set of "extensions" to the system provided
headers allowing for a better import into Swift. This may be in the form of API
Notes, module maps, wrapper types, or Swift `extension`s. This code may or may
not be fully inlined into the client code and thus be part of the platform ABI.

## Flags

> The compiler resources are controlled via the frontend flag `-resource-dir`.

Author: Oh? Hmm, I thought it was frontend.

> The compiler resources are controlled via the frontend flag `-resource-dir`.

Author: ```suggestion
The compiler resources are controlled via the driver flag `-resource-dir`.
```

> The compiler resources are controlled via the frontend flag `-resource-dir`.

Reviewer 4: It's both, the driver passes it to the frontend.

This allows the driver to select the correct location in most cases while
allowing the developer control to override the value if required. This should

> normally not be required to be altered as it is intrinsic to the compiler.

Reviewer 2: I heard you like bees.

Maybe something more like this? It's a bit more active-voiced.
```suggestion
Normally, you should not need to set this flag as the location of these files is intrinsic to the compiler.
```

The system headers are more interesting. Since this is C/C++ content, this is
actually consumed through the clang importer rather than the Swift compiler. The
Swift toolchain uses clang as the C/C++ compiler on all platforms as it is
embedded to generate inline FFI to enable seamless C/C++ bridging to Swift. The
flag used by clang is derived from the GCC toolchain, and is spelt `--sysroot`.

> Again, it is the responsibility of the Swift driver to identify the correct

Reviewer 2: If this isn't a run-on sentence, it's a very long one.

perhaps something more like this? Feel free to rework it a bit because I'm still not 100% happy with it.
```suggestion
The compiler driver is responsible for identifying the structure of the sysroot. When cross-compiling, there isn't a consistent location for these files, so the driver must expose an argument to specify where to find these files.
```

code against.

The compiler resources are content for the compiler that the compiler provides.
In the case of the Swift compiler, this includes the Swift shims. Whilst this is
a compiler resource, the packaging may not necessarily be part of the toolchain
due to interdependencies. Some of this content is required to process the
system headers themselves (e.g. clang's builtin resource headers).

The C/C++ system headers (and libraries) are what is traditionally called the
Unix "sysroot". On Darwin, this is the SDK which is shipped to the user, while
on Windows, this is called the Windows SDK, which is a separate installable
component.

The platform Swift modules provided by the SDK overlay overlays the sysroot or
(C/C++) SDK. This overlay is a set of "extensions" to the system provided
headers allowing for a better import into Swift. This may be in the form of API
Notes, module maps, wrapper types, or Swift `extension`s. This code may or may
not be fully inlined into the client code and thus be part of the platform ABI.

## Flags

The compiler resources are controlled via the driver flag `-resource-dir`.
This allows the driver to select the correct location in most cases while

> allowing the developer control to override the value if required. This should

Author: ```suggestion
allowing the developer control to override the value if required.
```

Normally, you should not need to set this flag as the location of these files is intrinsic to the compiler.

The system headers are more interesting. Since this is C/C++ content, this is
actually consumed through the clang importer rather than the Swift compiler. The
Swift toolchain uses clang as the C/C++ compiler on all platforms as it is
embedded to generate inline FFI to enable seamless C/C++ bridging to Swift. The
flag used by clang is derived from the GCC toolchain, and is spelt `--sysroot`.
The compiler driver is responsible for identifying the structure of the sysroot. When cross-compiling, there isn't a consistent location for these files, so the driver must expose an argument to specify where to find these files.
sysroot, although because there is no uniform installation location, this
parameter needs to be exposed to the user to allow selection, particularly in
the case of cross-compilation.

Currently, we do not have a good way to isolate the SDK overlay from the
remainder of the required content. On Darwin platforms, this content is shipped

> as part of the SDK. As a result the singular `-sdk` flag allows control over the

Reviewer 4: Note that the new swift-driver does not look in `-sdk` for the SDK overlay, only for the C/C++ headers, apple/swift-driver#1562.

location of the SDK overlay which is applied to the system SDK. By default, the
environment variable `SDKROOT` is used to seed the value of `-sdk`, though the
user may specify the value explicitly. Other platforms do not currently have a
flag to control this location and the toolchain defaults to a set of relative
paths to locate the content. This prevents cross-compilation as the included
content would be for a single platform.

## Solution

Generalising the above structure and sharing the common sharing gives way to the
following set of flags for cross-compilation:

1. `-target`: specifies the triple for the host
2. `-sysroot`: specifies the (C/C++) sysroot for the host platform content
3. `-sdk`: specifies the Swift SDK overlay for the host

The values for these may be defaulted by the driver on a per-platform basis.

The `-sysroot` flag identifies the location for the C/C++ headers and libraries required for compilation. This is primarily used by non-Darwin, non-Windows hosts as Darwin has its own SDK concept that allows for co-installation and Windows uses a different model which merges multiple locations in memory.

> The `-sdk` flag identifies the location of the "Swift SDK overlay", which provides the neecessary content for Swift compilation (including binary swiftmodules). This includes the standard library and the core libraries (dispatch, Foundation, and possibly XCTest - Windows isolates XCTest from the rest of the SDK). The Swift shims are also provided by this location as they are a dependency for properly processing the Swift core library.

Reviewer 5: ```suggestion
The `-sdk` flag identifies the location of the "Swift SDK overlay", which provides the necessary content for Swift compilation (including binary swiftmodules). This includes the standard library and the core libraries (dispatch, Foundation, and possibly XCTest - Windows isolates XCTest from the rest of the SDK). The Swift shims are also provided by this location as they are a dependency for properly processing the Swift core library.
```

# Cross-Compilation Model

## Components

When compiling Swift code, the compiler will consult three different sources of
inputs outside of the user provided code.

1. Compiler Resources
2. System (C/C++) Headers
3. Platform Swift Modules (SDK Overlay)

These pieces compose in a particular manner to build a system image to build
code against.

The compiler resources are content for the compiler that the compiler provides.

> In the case of the Swift compiler, this includes the Swift shims. Whilst this is

Author: I think I like the slightly more verbose version above for the compiler resources. I think that the system header version also suffers from nitpicking: I extracted a copy of some prebuilt version of the library - its technically an installation but the headers are not really part of the system per se.

Unix "sysroot". On Darwin, this is the SDK which is shipped to the user, while
on Windows, this is called the Windows SDK, which is a separate installable
component.

The platform Swift modules provided by the SDK overlay overlays the sysroot or
(C/C++) SDK. This overlay is a set of "extensions" to the system provided
headers allowing for a better import into Swift. This may be in the form of API
Notes, module maps, wrapper types, or Swift `extension`s. This code may or may
not be fully inlined into the client code and thus be part of the platform ABI.

## Flags

The compiler resources are controlled via the driver flag `-resource-dir`.
This allows the driver to select the correct location in most cases while
allowing the developer control to override the value if required.
Normally, you should not need to set this flag as the location of these files is intrinsic to the compiler.

The system headers are more interesting. Since this is C/C++ content, this is
actually consumed through the clang importer rather than the Swift compiler. The
Swift toolchain uses clang as the C/C++ compiler on all platforms as it is

> embedded to generate inline FFI to enable seamless C/C++ bridging to Swift. The

Reviewer 2: expand the first use of acronyms 

```suggestion
embedded to generate inline foreign function interfaces (FFI) to enable seamless C/C++ bridging to Swift. The
```

flag used by clang is derived from the GCC toolchain, and is spelt `--sysroot`.
The compiler driver is responsible for identifying the structure of the sysroot. When cross-compiling, there isn't a consistent location for these files, so the driver must expose an argument to specify where to find these files.

> sysroot, although because there is no uniform installation location, this

Reviewer 2: I think this sentence might be left-over from editing?

Author: Yeah, thanks!

version, SDK version, VCRuntime version). The `-sdk` flag is used to specify the
location of the SDK overlay which is applied to the system SDK. By default, the
environment variable `SDKROOT` is used to seed the value of `-sdk`, though the
user may specify the value explicitly. Other platforms do not currently have a
flag to control this location and the toolchain defaults to a set of relative
paths to locate the content. This prevents cross-compilation as the included
content would be for a single platform.

> [!NOTE]
> `-resource-dir` historically influenced the driver's search for the SDK content, locating the standard library relative to the resource directory. This behaviour is considered deprecated but remains for compatibility. The `-sdk` parameter is given precedence and is the preferred mechanism for controlling the behaviour of the driver to locate the SDK content.

## Solution

Generalising the above structure and sharing the common sharing gives way to the
following set of flags for cross-compilation:

1. `-target`: specifies the triple for the host
2. `-sysroot`: specifies the (C/C++) sysroot for the host platform content

> 3. `-sdk`: specifies the Swift SDK overlay for the host

Reviewer 2: Do we want to call it an SDK overlay if it is the Swift SDK for the platform?
It contains the C/C++ overlay for importing into Swift, sure, but it usually also contains Foundation, the standard library, etc... (arguably, Foundation could be its own library in its own directory structure and let the standard library exist in its own space, but that's an argument for a different venue.)

Author: I mean it still does overlay the platform SDK. I don't like Swift SDK because it can be confusing when you are talking about Darwin vs non-Darwin as the former will use a single unified SDK.
